src/range.rs

```diff
@@ -308,6 +308,154 @@ impl<V: Ord> Range<V> {
+        // right:  ∞-----
+        (Unbounded, Unbounded) => Ordering::Equal,
+        // left:     [---
+        // right:  ∞-----
+        (Included(_left), Unbounded) => Ordering::Greater,
+        // left:     ]---
+        // right:  ∞-----
+        (Excluded(_left), Unbounded) => Ordering::Greater,
+        // left:   ∞-----
+        // right:    [---
+        (Unbounded, Included(_right)) => Ordering::Less,
+        // left:   [----- OR [----- OR   [-----
+        // right:    [--- OR [----- OR [---
+        (Included(left), Included(right)) => left.partial_cmp(right)?,
+        (Excluded(left), Included(right)) => match left.partial_cmp(right)? {
+            // left:   ]-----
+            // right:    [---
+            Ordering::Less => Ordering::Less,
+            // left:   ]-----
+            // right:  [---
+            Ordering::Equal => Ordering::Greater,
+            // left:     ]---
+            // right:  [-----
+            Ordering::Greater => Ordering::Greater,
+        },
+        // left:   ∞-----
+        // right:    ]---
+        (Unbounded, Excluded(_right)) => Ordering::Less,
+        (Included(left), Excluded(right)) => match left.partial_cmp(right)? {
+            // left:   [-----
+            // right:    ]---
+            Ordering::Less => Ordering::Less,
+            // left:   [-----
+            // right:  ]---
+            Ordering::Equal => Ordering::Less,
+            // left:     [---
+            // right:  ]-----
+            Ordering::Greater => Ordering::Greater,
+        },
+        // left:   ]----- OR ]----- OR   ]---
+        // right:    ]--- OR ]----- OR ]-----
+        (Excluded(left), Excluded(right)) => left.partial_cmp(right)?,
+    })
+}
+
+/// Implementing `PartialOrd` for end `Bound` of an interval.
+///
+/// We flip the unbounded ranges from `-∞` to `∞`, while `V`-valued bounds checks remain the same.
+///
+/// Legend: `∞` is unbounded, `[1,2]` is `>=1,<=2`, `]1,2[` is `>1,<2`.
+///
+/// ```text
+/// left:   [--------∞
+/// right:  [-----]
+/// left is greater, since it starts earlier.
+///
+/// left:   [-----[
+/// right:  [-----]
+/// left is smaller, since it ends earlier.
+/// ```
+fn cmp_bounds_end<V: PartialOrd>(left: Bound<&V>, right: Bound<&V>) -> Option<Ordering> {
+    Some(match (left, right) {
+        // left:   -----∞
+        // right:  -----∞
+        (Unbounded, Unbounded) => Ordering::Equal,
+        // left:   ---]
+        // right:  -----∞
+        (Included(_left), Unbounded) => Ordering::Less,
+        // left:   ---[
+        // right:  -----∞
+        (Excluded(_left), Unbounded) => Ordering::Less,
+        // left:  -----∞
+        // right: ---]
+        (Unbounded, Included(_right)) => Ordering::Greater,
+        // left:   -----] OR -----] OR ---]
+        // right:    ---] OR -----] OR -----]
+        (Included(left), Included(right)) => left.partial_cmp(right)?,
+        (Excluded(left), Included(right)) => match left.partial_cmp(right)? {
```

Comment:
Would this be easier to read as `left.partial_cmp(right)?.then(Ordering::Less)`? It would be more concise, but not necessarily more readable.

Reply:
Personally, I like the explicitness instead of wondering what is the semantic of `then` for partial orders.

Reply:
I prefer as much explicitness as possible here, so that you can reason the correctness of each branch in isolation by looking at as few examples as possible.